File: StreamingBrowserElectron/tests/startphasentest.js
"use strict";
// Die Ladephasen beim Start einer Folge.
//
// Warum es diese Pruefung gibt. Der Ladebildschirm der App darf keinen
// erfundenen Fortschritt zeigen - er springt nur, wenn ein Schritt der
// Startkette wirklich hinter ihr liegt. Genau das laesst sich ohne Geraet
// pruefen, und genau das ist die Zusage, die hier festgehalten wird:
//
//   1. Der Balken geht nie rueckwaerts. Die Kette meldet Schritte mehrfach -
//      jeder Takt des Autostarts sieht denselben Zustand wieder.
//   2. Ohne gespeicherten Stand gibt es den Schritt "Zur gespeicherten Stelle"
//      gar nicht; er stuende sonst als Schritt da, den niemand je meldet.
//   3. Jeder Schritt hat seine eigene Frist, und ueber allem steht ein Deckel.
//      Beides muss greifen, sonst haengt ein Ladebildschirm fuer immer.
//   4. Das Modell, das Android beim Start abholt, traegt alles, was die App
//      zum Zeichnen braucht - Namen, Beschriftungen, Anteile, Fristen und die
//      Fehlertexte. Fehlt dort etwas, faellt es auf dem Geraet auf und nicht
//      hier, und dann steht die Tabelle zweimal.

const startphasen = require("../src/startphasen");
const autostart = require("../src/watchparty-autostart");

const pruefungen = [];
const pruefe = (n, b, d) => {
  pruefungen.push(Boolean(b));
  console.log(`${b ? "OK  " : "FAIL"}  ${n}${d ? "   -> " + d : ""}`);
};

/* ------------------------------------------------------ Die Reihenfolge */

const namen = startphasen.PHASEN.map((phase) => phase.name);
pruefe("Die Schritte stehen in der Reihenfolge der Startkette",
  namen.join(",") === "seite,hoster,spieler,stelle,vollbild,laeuft", namen.join(","));

pruefe("Der Balken waechst mit jedem Schritt",
  startphasen.PHASEN.every((phase, i) => i === 0 || phase.anteil > startphasen.PHASEN[i - 1].anteil));

pruefe("Der letzte Schritt ist der volle Balken",
  startphasen.anteil(startphasen.LETZTE) === 1);

pruefe("Der erste Schritt faengt nicht bei null an - etwas geschieht ja schon",
  startphasen.anteil(startphasen.ERSTE) > 0 && startphasen.anteil(startphasen.ERSTE) < 0.2);

pruefe("Jeder Schritt hat eine Beschriftung",
  startphasen.PHASEN.every((phase) => typeof phase.text === "string" && phase.text.length > 3));

/* ------------------------------------------------- Der Stand entscheidet */

const mitStand = startphasen.phasen({ stelle: 182.7 }).map((p) => p.name);
const ohneStand = startphasen.phasen({ stelle: 0 }).map((p) => p.name);
pruefe("Mit gespeichertem Stand gehoert der Sprung dazu", mitStand.includes("stelle"));
pruefe("Ohne gespeicherten Stand faellt er heraus", !ohneStand.includes("stelle"),
  ohneStand.join(","));
pruefe("Sonst sind es dieselben Schritte",
  ohneStand.join(",") === mitStand.filter((n) => n !== "stelle").join(","));

/* --------------------------------------------------- Nur nach vorn */

{
  const lauf = startphasen.starten({ titel: "Attack on Titan · Staffel 2 Folge 5", stelle: 182.7, jetzt: 1000 });
  pruefe("Ein Start faengt beim ersten Schritt an", lauf.phase === "seite");

  const a = startphasen.melden(lauf, "spieler", 2000);
  pruefe("Ein Sprung nach vorn zaehlt", a.geaendert && lauf.phase === "spieler");

  const b = startphasen.melden(lauf, "hoster", 3000);
  pruefe("Ein Schritt hinter dem Stand wird verworfen",
    !b.geaendert && lauf.phase === "spieler", lauf.phase);

  const c = startphasen.melden(lauf, "spieler", 4000);
  pruefe("Derselbe Schritt noch einmal aendert nichts", !c.geaendert && lauf.seit === 2000);

  const d = startphasen.melden(lauf, "kennt-keiner", 5000);
  pruefe("Ein unbekannter Name aendert nichts", !d.geaendert && lauf.phase === "spieler");

  const e = startphasen.melden(lauf, "laeuft", 6000);
  pruefe("Der letzte Schritt macht den Start fertig", e.fertig && lauf.fertig);
  pruefe("Und faerbt den Balken voll", e.anteil === 1, String(e.anteil));
}

/* --------------------------------------------------------- Die Fristen */

{
  const lauf = startphasen.starten({ titel: "T", stelle: 0, jetzt: 0 });
  const frist = startphasen.PHASEN[0].fristMs;

  pruefe("Kurz vor der Frist wird gewartet",
    startphasen.pruefen(lauf, frist - 1).tun === "warten");

  const abgelaufen = startphasen.pruefen(lauf, frist);
  pruefe("Mit der Frist ist Schluss", abgelaufen.tun === "fehler" && abgelaufen.grund === "seite",
    abgelaufen.grund);
  pruefe("Und der Zuschauer bekommt einen Satz, mit dem er etwas anfangen kann",
    /Internetverbindung/i.test(abgelaufen.text), abgelaufen.text);
}

{
  // Jeder Schritt kommt gerade noch rechtzeitig - und trotzdem ist irgendwann
  // Schluss. Ohne den Deckel liesse sich ein Ladebildschirm beliebig lange
  // offenhalten, indem jede Phase kurz vor ihrer Frist weiterspringt.
  const lauf = startphasen.starten({ titel: "T", stelle: 5, jetzt: 0 });
  let zeit = 0;
  for (const name of ["hoster", "spieler"]) {
    // Immer eine Sekunde vor der Frist des laufenden Schrittes melden: jeder
    // einzelne Schritt bleibt damit im Rahmen.
    zeit += startphasen.PHASEN.find((p) => p.name === lauf.phase).fristMs - 1000;
    startphasen.melden(lauf, name, zeit);
    pruefe("Rechtzeitig gemeldet, also weiter (" + name + ")",
      startphasen.pruefen(lauf, zeit + 10).tun === "warten");
  }
  pruefe("Zusammen sind sie laenger als der Deckel", zeit > startphasen.GESAMT_FRIST_MS - 5000,
    zeit + " ms");
  const spaet = startphasen.pruefen(lauf, startphasen.GESAMT_FRIST_MS);
  pruefe("Der Deckel greift auch dann", spaet.tun === "fehler" && spaet.grund === "gesamt",
    spaet.grund);
}

{
  const lauf = startphasen.starten({ titel: "T", stelle: 0, jetzt: 0 });
  startphasen.melden(lauf, "laeuft", 10);
  pruefe("Ein fertiger Start laeuft in keine Frist",
    startphasen.pruefen(lauf, startphasen.GESAMT_FRIST_MS * 5).tun === "warten");
}

/* -------------------------------------------- Das Modell fuer das Geraet */

{
  const modell = startphasen.modell();
  pruefe("Das Modell traegt alle Schritte", modell.phasen.length === startphasen.PHASEN.length);
  pruefe("Mit Namen, Text, Anteil und Frist", modell.phasen.every((phase) => (
    typeof phase.name === "string" && phase.name.length > 0
      && typeof phase.text === "string" && phase.text.length > 0
      && typeof phase.anteil === "number"
      && typeof phase.fristMs === "number"
  )));
  pruefe("Es nennt Anfang und Ende",
    modell.erste === startphasen.ERSTE && modell.letzte === startphasen.LETZTE);
  pruefe("Und den Deckel", modell.gesamtFristMs === startphasen.GESAMT_FRIST_MS);
  pruefe("Zu jedem Schritt steht ein Fehlertext",
    startphasen.PHASEN.every((phase) => (modell.fehlertexte[phase.name] || "").length > 5));
  pruefe("Dazu der Deckel und der Abbruch",
    (modell.fehlertexte.gesamt || "").length > 5 && (modell.fehlertexte.abgebrochen || "").length > 5);
  pruefe("Und ein Satz fuer alles Uebrige", (modell.fehlertexte[""] || "").length > 5);
  pruefe("Das Modell ist eine Kopie - wer daran dreht, dreht nicht an der Tabelle",
    (() => {
      modell.phasen[0].text = "kaputt";
      return startphasen.beschriftung(startphasen.ERSTE) !== "kaputt";
    })());
}

/* ------------------------------- Die Zwischenmeldungen aus dem Player */

// Sie sind der Grund, warum der Balken zwischen "Player gefunden" und "es
// laeuft" ueberhaupt etwas zu melden hat. Ohne sie stuende er dort zehn
// Sekunden still - und genau dann faengt jemand an, Zeit hochzuzaehlen.
{
  const skript = autostart.startScript("auftrag-1", { videoTime: 12, timestamp: Date.now(), playing: true, hatUhr: true }, { playing: true });
  pruefe("Das Startskript meldet den gefundenen Rahmen",
    skript.includes(`meldePhase(${JSON.stringify(autostart.PHASE_RAHMEN)})`));
  pruefe("Und die geladene Quelle",
    skript.includes(`meldePhase(${JSON.stringify(autostart.PHASE_QUELLE)})`));
  pruefe("Ueber denselben Kanal wie der Bericht - die Konsole",
    skript.includes(`console.log("${autostart.MELDE_PHASE}"`));
  pruefe("Die Meldung faengt mit dem gemeinsamen Praefix an - sonst hoert Java gar nicht hin",
    autostart.MELDE_PHASE.startsWith("__elfix:wp:"), autostart.MELDE_PHASE);
  pruefe("Und laesst sich wieder auseinandernehmen",
    autostart.phaseLesen(autostart.MELDE_PHASE + "quelle") === "quelle");
  pruefe("Was keine ist, ist keine",
    autostart.phaseLesen(autostart.MELDE_START + "{}") === "");
  pruefe("Der gemeldete Rahmenname ist einer der Schritte",
    namen.includes(autostart.PHASE_RAHMEN), autostart.PHASE_RAHMEN);
}

const fehlgeschlagen = pruefungen.filter((ok) => !ok).length;
console.log(`\n${pruefungen.length - fehlgeschlagen}/${pruefungen.length} bestanden`);
process.exit(fehlgeschlagen ? 1 : 0);

File: android/app/src/test/java/local/elflix/android/StartphasenTest.java
package local.elflix.android;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import org.json.JSONObject;
import org.junit.Test;

/**
 * Die Ladephasen, so wie das Geraet sie liest.
 *
 * <h2>Was hier geprueft wird - und was ausdruecklich nicht</h2>
 *
 * <p>Nicht die Tabelle. Namen, Beschriftungen, Anteile und Fristen gehoeren
 * dem geteilten Modul {@code src/startphasen.js}; sie stehen dort einmal und
 * werden dort geprueft ({@code tests/startphasentest.js}). Ein zweiter Satz
 * Zahlen hier waere genau die zweite Wahrheit, die dieser Aufbau vermeiden
 * soll.
 *
 * <p>Geprueft wird der Weg dazwischen: dass {@link Startphasen} das Modell
 * vollstaendig uebernimmt, dass es ohne Modell ehrlich {@code false} meldet -
 * dann gibt es keinen Vorhang, und der Start laeuft wie vorher -, und dass die
 * beiden Rechnungen des Vorhangs stimmen. Es sind genau zwei: ist eine
 * gemeldete Phase weiter als die stehende, und ist eine Frist abgelaufen.
 *
 * <p>Das Modell im Text unten ist eine gekuerzte Nachbildung mit anderen
 * Zahlen als das echte. Das ist Absicht: eine Pruefung, die dieselben Werte
 * nennt wie die Tabelle, prueft nur, dass jemand richtig abgeschrieben hat.
 */
public class StartphasenTest {

    private static final String MODELL = "{"
        + "\"phasen\":["
        + "{\"name\":\"seite\",\"text\":\"Folge wird geöffnet\",\"anteil\":0.1,\"fristMs\":5000},"
        + "{\"name\":\"hoster\",\"text\":\"Player wird geladen\",\"anteil\":0.4,\"fristMs\":3000},"
        + "{\"name\":\"laeuft\",\"text\":\"Wiedergabe läuft\",\"anteil\":1,\"fristMs\":0}"
        + "],"
        + "\"erste\":\"seite\",\"letzte\":\"laeuft\",\"gesamtFristMs\":9000,"
        + "\"fehlertexte\":{\"seite\":\"Die Folgenseite lädt nicht.\","
        + "\"hoster\":\"Der Hoster hat keinen Player geliefert.\","
        + "\"gesamt\":\"Der Start dauert zu lange.\","
        + "\"\":\"Die Folge konnte nicht gestartet werden.\"}"
        + "}";

    private static Startphasen geladen() throws Exception {
        Startphasen phasen = new Startphasen(null);
        phasen.uebernehmen(new JSONObject(MODELL));
        return phasen;
    }

    /* -------------------------------------------------------- Ohne Modell */

    @Test
    public void ohneModellGibtEsKeinenVorhang() {
        Startphasen phasen = new Startphasen(null);
        assertFalse("ohne Tabelle ist nichts bereit", phasen.istBereit());
        assertEquals("", phasen.erste());
        assertEquals("", phasen.letzte());
        assertEquals(-1, phasen.nummer("seite"));
    }

    @Test
    public void einLeeresModellWirdNichtUebernommen() throws Exception {
        Startphasen phasen = geladen();
        phasen.uebernehmen(new JSONObject("{\"phasen\":[]}"));
        assertTrue("die geladene Tabelle bleibt stehen", phasen.istBereit());
        assertEquals("seite", phasen.erste());
    }

    /* ---------------------------------------------------- Die Uebernahme */

    @Test
    public void dasModellWirdVollstaendigUebernommen() throws Exception {
        Startphasen phasen = geladen();
        assertTrue(phasen.istBereit());
        assertEquals("seite", phasen.erste());
        assertEquals("laeuft", phasen.letzte());
        assertEquals(9000L, phasen.gesamtFristMs());
        assertEquals("Player wird geladen", phasen.beschriftung("hoster"));
        assertEquals(0.4, phasen.anteil("hoster"), 0.0001);
        assertEquals(3000L, phasen.phase("hoster").fristMs);
    }

    @Test
    public void unbekannteNamenAntwortenLeerUndNichtMitEinerAusnahme() throws Exception {
        Startphasen phasen = geladen();
        assertEquals(-1, phasen.nummer("gibt-es-nicht"));
        assertEquals("", phasen.beschriftung("gibt-es-nicht"));
        assertEquals(0.0, phasen.anteil("gibt-es-nicht"), 0.0001);
        assertEquals(null, phasen.phase("gibt-es-nicht"));
    }

    @Test
    public void jederSchrittHatSeinenEigenenFehlertext() throws Exception {
        Startphasen phasen = geladen();
        assertEquals("Der Hoster hat keinen Player geliefert.", phasen.fehlertext("hoster"));
        assertEquals("Der Start dauert zu lange.", phasen.fehlertext("gesamt"));
    }

    @Test
    public void wasKeinenEigenenHatBekommtDenAllgemeinen() throws Exception {
        Startphasen phasen = geladen();
        assertEquals("Die Folge konnte nicht gestartet werden.", phasen.fehlertext("laeuft"));
        assertEquals("Die Folge konnte nicht gestartet werden.", phasen.fehlertext("was-auch-immer"));
        assertEquals("Die Folge konnte nicht gestartet werden.", phasen.fehlertext(null));
    }

    /* ------------------------------------------------ Nur nach vorn */

    // Die Kette meldet Schritte mehrfach - jeder Takt des Autostarts sieht
    // denselben Zustand wieder. Ein Balken, der dabei zurueckspringt, sieht aus
    // wie ein Fehler. Der Vorhang vergleicht dafuer die Nummern, und genau das
    // steht hier.
    @Test
    public void schritteLassenSichDerReiheNachVergleichen() throws Exception {
        Startphasen phasen = geladen();
        assertTrue(phasen.nummer("hoster") > phasen.nummer("seite"));
        assertTrue(phasen.nummer("laeuft") > phasen.nummer("hoster"));
        assertFalse("rueckwaerts ist keine Meldung",
            phasen.nummer("seite") > phasen.nummer("hoster"));
        assertFalse("und ein unbekannter Name erst recht nicht",
            phasen.nummer("gibt-es-nicht") > phasen.nummer("seite"));
    }

    /* ------------------------------------------------------- Die Fristen */

    @Test
    public void kurzVorDerFristWirdGewartet() throws Exception {
        Startphasen phasen = geladen();
        assertEquals("", Startvorhang.abgelaufen(phasen, "seite", 0, 0, 4999));
    }

    @Test
    public void mitDerFristDesSchrittesIstSchluss() throws Exception {
        Startphasen phasen = geladen();
        assertEquals("seite", Startvorhang.abgelaufen(phasen, "seite", 0, 0, 5000));
    }

    @Test
    public void jederSchrittHatSeineEigeneGeduld() throws Exception {
        Startphasen phasen = geladen();
        // Bei 4000 ms in der Phase waere "seite" noch geduldig, "hoster" nicht.
        assertEquals("", Startvorhang.abgelaufen(phasen, "seite", 0, 0, 4000));
        assertEquals("hoster", Startvorhang.abgelaufen(phasen, "hoster", 0, 0, 4000));
    }

    @Test
    public void derDeckelGehtVor() throws Exception {
        Startphasen phasen = geladen();
        // Der Schritt selbst laeuft erst seit einer Sekunde und waere geduldig -
        // aber der ganze Start dauert nun schon neun.
        assertEquals("gesamt", Startvorhang.abgelaufen(phasen, "hoster", 0, 8000, 9000));
    }

    @Test
    public void einSchrittOhneFristLaeuftInKeine() throws Exception {
        Startphasen phasen = geladen();
        assertEquals("", Startvorhang.abgelaufen(phasen, "laeuft", 8000, 8000, 8999));
    }

    @Test
    public void ohneTabelleWirdNichtsFuerAbgelaufenErklaert() {
        assertEquals("", Startvorhang.abgelaufen(null, "seite", 0, 0, 999999));
        assertEquals("", Startvorhang.abgelaufen(new Startphasen(null), "seite", 0, 0, 999999));
    }

    /* --------------------------------------------------- Die Zeitangabe */

    // Nur fuer die Anzeige: "Zur gespeicherten Stelle (3:03)". Sie steht hier,
    // weil eine falsch gerundete Minute im Ladebildschirm sofort auffaellt und
    // sich sonst niemand darum kuemmert.
    @Test
    public void derGespeicherteStandStehtLesbarDa() {
        assertEquals("0:00", Startvorhang.zeitText(0));
        assertEquals("0:07", Startvorhang.zeitText(7.4));
        assertEquals("3:03", Startvorhang.zeitText(182.7));
        assertEquals("23:40", Startvorhang.zeitText(1420));
        assertEquals("1:00:00", Startvorhang.zeitText(3600));
        assertEquals("1:02:03", Startvorhang.zeitText(3723));
    }
}
